packages/grafana-runtime/src/utils/DataSourceWithBackend.ts

```diff
@@ -132,8 +136,44 @@ class DataSourceWithBackend<
         requestId,
       })
       .pipe(
-        map((rsp) => {
-          return toDataQueryResponse(rsp, queries as DataQuery[]);
+        switchMap((raw) => {
+          const rsp = toDataQueryResponse(raw, queries as DataQuery[]);
+          // Check if any response should subscribe to a live stream
+          if (rsp.data?.length && rsp.data.find((f: DataFrame) => f.meta?.channel)) {
+            const buffer: StreamingFrameOptions = {
+              maxLength: request.maxDataPoints ?? 500,
+            };
+
+            // For recent queries, clamp to the current time range
+            if (request.rangeRaw?.to === 'now') {
+              buffer.maxDelta = request.range.to.valueOf() - request.range.from.valueOf();
+            }
+
+            const staticdata: DataFrame[] = [];
+            const streams: Array<Observable<DataQueryResponse>> = [];
+            for (const frame of rsp.data) {
+              const addr = parseLiveChannelAddress(frame.meta?.channel);
+              if (addr) {
+                streams.push(
+                  getLiveDataStream({
+                    addr,
+                    buffer,
+                    frame: frame as DataFrame,
+                  })
+                );
+              } else {
+                staticdata.push(frame);
+              }
+            }
+            if (staticdata.length) {
+              streams.push(of({ ...rsp, data: staticdata }));
+            }
+            if (streams.length === 1) {
+              return streams[0]; // avoid merge wrapper
+            }
+            return merge(...streams);
+          }
+          return of(rsp);
```

Comment:
Is there anyway we can pull this apart so it becomes testable? And maybe add tests too?